adds a mechanism to inject additional steps into mission loading

Azaezel opened the pull request:
leverages the EventManager and ScriptMsgListener() classes to set up a third mission load stage triggered by the following flow: 
function <module>::onLoadMap(%this) starts an execution chain that leads to <module>::finishMapLoad().
each  <module>::finishMapLoad() MUST contain the line
     Core_ClientServer.GetEventManager().postEvent( "mapLoadComplete" );
once all have called back that they have finished their tasks, players finish loading into a hosted mission

Azaezel committed:
adds a mechanism to inject additional steps into mission loading
leverages the EventManager and ScriptMsgListener() classes to set up a third mission load stage triggered by the following flow:
function <module>::onLoadMap(%this) starts an execution chain that leads to <module>::finishMapLoad()
each  <module>::finishMapLoad() MUST contain the line
     Core_ClientServer.GetEventManager().postEvent( "mapLoadComplete" );
once all have called back that they have finished thier tasks, players finish loading into a hosted mission

Azaezel commented:
Will note: Never a real fan of musts, so by all means, open to alternative suggestions

dottools commented:
You know what this means. Soon we'll need ordered dependency mission loading callbacks. Followed by optional retries on failure. Eventually we'll get to systemd level craziness and we'll just have to curl up and cry ourselves to sleep each night.... ;)

Azaezel commented:
weeeell... as far as that goes, module order dependency's already a thing. and the retry on a fail bit is one of the reasons for cooking this up to begin with. was working on a procedural generation module with a potential for a failed outcome where it retries. which could take a bit. hence the 'and get back to me when you're done' injection via listener. so already a good 50-75% there on those fronts. Though higher order interleaving is likely beyond scope till 4.1 proper.

Azaezel committed:
adds a fail state method+event to the load chain
at time of writing:
    Core_ClientServer.GetEventManager().postEvent( "mapLoadFail", false );
kicks players and closes the server
    Core_ClientServer.GetEventManager().postEvent( "mapLoadFail", false );
proceeds to continue loading reguardless

Azaezel committed:
refactor, with the following wrappers to keep in mind:
Core_ClientServer.clearLoadStatus();
Core_ClientServer.inishMapLoad();
Core_ClientServer.FailMapLoad( %moduleName, %isFine);
Of special note: the postevent method must only take one entry, so we store off  Core_ClientServer.failedModuleName = %moduleName; priorto triggering the event so that the failing module can be reported.
